htlcswitch/circuit_map.go

```diff
@@ -0,0 +1,848 @@
+				circuits = append(circuits, circuit)
+			}
+		}
+	}
+
+	return circuits
+}
+
+// CommitCircuits accepts any number of circuits and persistently adds them to
+// the switch's circuit map. The method returns a list of circuits that had not
+// been seen prior by the switch. A link should only forward HTLCs corresponding
+// to the returned circuits to the switch.
+//
+// NOTE: This method uses batched writes to improve performance, gains will only
+// be realized if it is called concurrently from separate goroutines.
+func (cm *circuitMap) CommitCircuits(circuits ...*PaymentCircuit) (
+	*CircuitFwdActions, error) {
+
+	actions := &CircuitFwdActions{}
+
+	// If an empty list was passed, return early to avoid grabbing the lock.
+	if len(circuits) == 0 {
+		return actions, nil
+	}
+
+	// First, we reconcile the provided circuits with our set of pending
+	// circuits to construct a set of new circuits that need to be written
+	// to disk. The circuit's pointer is stored so that we only permit this
+	// exact circuit to be forwarded through the switch. If a circuit is
+	// already pending, the htlc will be reforwarded by the switch.
+	cm.mtx.Lock()
+	var adds, drops, fails, addFails []*PaymentCircuit
+	for _, circuit := range circuits {
+		inKey := circuit.InKey()
+		if foundCircuit, ok := cm.pending[inKey]; ok {
+			switch {
+
+			// This circuit has a keystone, it's waiting for a
+			// response from the remote peer on the outgoing link.
+			// Drop it like it's hot, ensure duplicates get caught.
+			case foundCircuit.HasKeystone():
+				drops = append(drops, circuit)
+
+			// If no keystone is set and the switch has not been
+			// restarted, the corresponding packet should still be
+			// in the outgoing link's mailbox. It will be delivered
+			// if it comes online before the switch goes down.
+			//
+			// NOTE: Dropping here prevents a flapping, incoming
+			// link from failing a duplicate add while it is still
+			// in the server's memory mailboxes.
+			case !foundCircuit.LoadedFromDisk:
+				drops = append(drops, circuit)
+
+			// Otherwise, the in-mem packet has been lost due to a
+			// restart. It is now safe to send back a failure along
+			// the incoming link. The incoming link should be able
+			// detect and ignore duplicate packets of this type.
+			default:
+				fails = append(fails, circuit)
+				addFails = append(addFails, circuit)
+			}
+
+			continue
+		}
+
+		cm.pending[inKey] = circuit
+		adds = append(adds, circuit)
+		addFails = append(addFails, circuit)
+	}
+	cm.mtx.Unlock()
+
+	// If all circuits are dropped or failed, we are done.
+	if len(adds) == 0 {
+		actions.Drops = drops
+		actions.Fails = fails
+		return actions, nil
+	}
+
+	// Now, optimistically serialize the circuits to add.
+	var bs = make([]bytes.Buffer, len(adds))
+	for i, circuit := range adds {
+		if err := circuit.Encode(&bs[i]); err != nil {
+			actions.Drops = drops
+			actions.Fails = addFails
+			return actions, err
```

Comment:
hhm, comment why this is returned and not just `nil, err`?

Reply:
This allows the caller to fail back any htlcs and log the error. In the future, we will implement more of a "safe shutdown" for the switch. So the error will recorded and the switch will try to flush any packets from mem before shutting down completely